File: bst/BinaryTreeTraversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// Definition of the Node class for the binary tree
class Node {
    int data;
    Node left, right;

    public Node(int item) {
        data = item;
        left = right = null;
    }
}

// Binary Tree class to perform different tree traversals
public class BinaryTreeTraversal {
    Node root;

    // Inorder Traversal (Left -> Root -> Right)
    void inorderTraversal(Node node) {
        if (node == null) {
            return;
        }
        inorderTraversal(node.left);  // Visit left subtree
        System.out.print(node.data + " ");  // Visit root node
        inorderTraversal(node.right);  // Visit right subtree
    }

    // Preorder Traversal (Root -> Left -> Right)
    void preorderTraversal(Node node) {
        if (node == null) {
            return;
        }
        System.out.print(node.data + " ");  // Visit root node
        preorderTraversal(node.left);  // Visit left subtree
        preorderTraversal(node.right);  // Visit right subtree
    }

    // Postorder Traversal (Left -> Right -> Root)
    void postorderTraversal(Node node) {
        if (node == null) {
            return;
        }
        postorderTraversal(node.left);  // Visit left subtree
        postorderTraversal(node.right);  // Visit right subtree
        System.out.print(node.data + " ");  // Visit root node
    }

    // Level Order Traversal (Breadth-First)
    void levelOrderTraversal() {
        if (root == null) {
            return;
        }

        Queue<Node> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            Node currentNode = queue.poll();
            System.out.print(currentNode.data + " ");  // Visit node

            // Add left child to queue
            if (currentNode.left != null) {
                queue.add(currentNode.left);
            }

            // Add right child to queue
            if (currentNode.right != null) {
                queue.add(currentNode.right);
            }
        }
    }

    // OR Level Order Traversal (Breadth-First)
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> currentLevel = new ArrayList<>();
            for (int i = 0; i < levelSize; i++) {
                Node node = queue.poll();
                currentLevel.add(node.data);
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            result.add(currentLevel);
        }
        return result;
    }

    public static void main(String[] args) {
        BinaryTreeTraversal tree = new BinaryTreeTraversal();

        // Construct the binary tree
        tree.root = new Node(50);
        tree.root.left = new Node(30);
        tree.root.right = new Node(70);
        tree.root.left.left = new Node(20);
        tree.root.left.right = new Node(40);
        tree.root.right.left = new Node(60);
        tree.root.right.right = new Node(80);

        System.out.println("Inorder traversal (Left -> Root -> Right): ");
        tree.inorderTraversal(tree.root);  // Expected: 20 30 40 50 60 70 80
        System.out.println("\n");

        System.out.println("Preorder traversal (Root -> Left -> Right), DFS: ");
        tree.preorderTraversal(tree.root);  // Expected: 50 30 20 40 70 60 80
        System.out.println("\n");

        System.out.println("Postorder traversal (Left -> Right -> Root): ");
        tree.postorderTraversal(tree.root);  // Expected: 20 40 30 60 80 70 50
        System.out.println("\n");

        System.out.println("Level order traversal (Breadth-First): ");
        tree.levelOrderTraversal();  // Expected: 50 30 70 20 40 60 80
        System.out.println();
    }
}

File: bst/BinaryTreeTraversal.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTreeTraversal:
    def inorderTraversal(self, root):
        result = []
        self.inorder(root, result)
        return result

    def inorder(self, root, result):
        if root:
            self.inorder(root.left, result)
            result.append(root.val)
            self.inorder(root.right, result)

    def preorderTraversal(self, root):
        result = []
        self.preorder(root, result)
        return result

    def preorder(self, root, result):
        if root:
            result.append(root.val)
            self.preorder(root.left, result)
            self.preorder(root.right, result)

    def postorderTraversal(self, root):
        result = []
        self.postorder(root, result)
        return result

    def postorder(self, root, result):
        if root:
            self.postorder(root.left, result)
            self.postorder(root.right, result)
            result.append(root.val)

    def levelOrder(self, root):
        result = []
        if not root:
            return result
        queue = deque([root])
        while queue:
            level = []
            for _ in range(len(queue)):
                node = queue.popleft()
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            result.append(level)
        return result

# Example usage
root = TreeNode(50)
root.left = TreeNode(30)
root.right = TreeNode(70)
root.left.left = TreeNode(20)
root.left.right = TreeNode(40)
root.right.left = TreeNode(60)
root.right.right = TreeNode(80)

solution = BinaryTreeTraversal()
print("Inorder traversal:", solution.inorderTraversal(root))
print("Preorder traversal:", solution.preorderTraversal(root))
print("Postorder traversal:", solution.postorderTraversal(root))
print("Level order traversal:", solution.levelOrder(root))